### PR TITLE
[Multi-User Part 2]: Add login pages and gate access to application behind login wall

### PR DESCRIPTION
**Pull request by sabaimran**
# Incoming
- Make most routes conditional on authentication *if anonymous mode is not enabled*. If anonymous mode is enabled, it scaffolds a default user and uses that for all application interactions.
- Add a basic login page and add routes for redirecting the user if logged in

### CODE REVIEW

File: src/khoj/configure.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ async def authenticate(self, request):
             user = await self.khojuser_manager.filter(email=current_user.get("email")).afirst()
             if user:
```
Comment (sabaimran) [inline src/khoj/configure.py:68]: In the `authenticate` method of the middleware, we should also check whether the `current_user`'s session object has an `exp` timestamp that exceeds the current time.

File: src/khoj/routers/api.py [hunk 2]
```diff
@@ -127,13 +127,15 @@ def _initialize_config():
             state.processor_config = configure_processor(state.config.processor)
 
     @api.get("/config/data", response_model=FullConfig)
+    @requires(["authenticated"], redirect="login_page")
```
Comment (debanjum) [inline src/khoj/routers/api.py:130]: These API endpoints (apart from Web client endpoints) have been updated in next PRs, to not redirect to login page but return unauthorized http status code in response. So no change required here

File: src/khoj/routers/indexer.py [hunk 3]
```diff
@@ -156,7 +157,6 @@ async def update(
         host=host,
     )
 
-    logger.info(f"📪 Content index updated via API call by {client} client")
```
Comment (debanjum) [inline src/khoj/routers/indexer.py:159]: Shouldn't this stay here?

File: src/khoj/main.py [hunk 4]
```diff
@@ -14,13 +14,16 @@
...
 from fastapi import FastAPI
 from fastapi.middleware.cors import CORSMiddleware
+from rich.logging import RichHandler
 import schedule
-import django
 
 from fastapi.staticfiles import StaticFiles
 from rich.logging import RichHandler
+import schedule
+
```
Comment (debanjum) [inline src/khoj/main.py:26]: Why are these showing up as new imports? None of the new code in this file seems to need these packages

File: src/khoj/interface/web/login.html [hunk 5]
```diff
@@ -0,0 +1,197 @@
+<html>
+    <head>
+        <meta charset="utf-8">
+        <meta name="viewport" content="width=device-width, initial-scale=1.0 maximum-scale=1.0">
+        <title>Khoj - Search</title>
```
Comment (debanjum) [inline src/khoj/interface/web/login.html:5]: Should the title say "Khoj - Config"?

```suggestion
        <title>Khoj - Config</title>
```

File: src/khoj/interface/web/login.html [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                    }),
+                    headers: {
+                        "Content-Type": "application/json"
+                    }
+                }).then(function(response) {
+                    return response.json();
+                }).then(function(data) {
+                    console.log(data);
+                    if (data.user != null) {
+                        document.getElementById("khoj-banner").innerHTML = "Thanks for signing up. We'll be in touch soon! 🚀";
```
Comment (debanjum) [inline src/khoj/interface/web/login.html:185]: Is this the text to be shown post signup?

**Review by debanjum — APPROVED** [review]
Looks good! Some of the changes suggested for this PR may already be addressed in next PRs for the rearchitecture work. Feel free to ignore them in this PR

File: src/khoj/interface/web/login.html (hunk 6 above)
Comment (sabaimran) [reply src/khoj/interface/web/login.html:185]: Oh, that's only if it's a demo instance. We actually don't need this here -- Probably won't have a demo instance + sign-in.

**Pull request merged**